File: 16-3sum-closest/16-3sum-closest.java
// ¡¡¡gesah für binäre Suche
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int ans = nums[0] + nums[1] + nums[2];
        int n = nums.length;
        Arrays.sort(nums);
        for (int i = 0; i < n - 2 && ans != target; i ++) {
            for (int j = i + 1; j < n - 1; j ++) {
                int complement = target - nums[i] - nums[j];
                int idx = Arrays.binarySearch(nums, j + 1, n - 1, complement);
                int hi = idx >= 0 ? idx : -idx;
                int lo = hi - 1;
                if (hi < n && Math.abs(complement - nums[hi]) < Math.abs(target - ans))
                    ans = nums[i] + nums[j] + nums[hi];
                if (lo < n && Math.abs(complement - nums[lo]) < Math.abs(target - ans))
                    ans = nums[i] + nums[j] + nums[lo];
            }
        }
        return ans;
    }
}

File: 16-3sum-closest/16-3sum-closest.py
# Q16m, runtime 12.63%
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        ans = sum(nums[:3])  #!!! caution
        # iter i to find [i, j, k]
        for i in range(len(nums) - 2):
            l, r = i + 1, len(nums) - 1
            while l < r:
                isum = nums[i] + nums[l] + nums[r]
                if isum == target:
                    return target
                elif isum < target:
                    l += 1
                else:
                    r -= 1
                ans = ans if abs(ans - target) < abs(isum - target) else isum
        return ans
    
'''
[0,1,2]
0
'''
